Utility function to get a setup & cleanup function for mapping each partition

squito opened the pull request:
Often when mapping some RDD, you want to do a bit of setup before processing each partition, followed by cleanup at the end of the partition; this adds utility functions to make that easier.

I felt that this is worth including because its a little tricky to get right -- I needed to add a "CleanupIterator", and I have an example in the unit tests of how this fails w/out it.  OTOH, I wasn't sure if this necessarily belongs in the spark api itself (eg., do we add a version of foreach with this also?)

We find it a useful addition, and so thought others might also ...

Commit:
sketch of how to do mapParitionsWithSetupAndCleanup

Commit:
add tests & docs

JoshRosen commented on core/src/test/scala/spark/RDDSuite.scala:
```diff
@@ -4,6 +4,8 @@ import scala.collection.mutable.HashMap
 import spark.SparkContext._
 import spark.rdd.{CoalescedRDD, PartitionPruningRDD}
+import spark.RDD.PartitionMapper
+import collection._
```
This style of import doesn't match the convention used in the rest of the codebase.  It should probably be replaced with an import of `scala.collection.mutable.Set`, since that appears to be the only class that it's importing.

JoshRosen commented on core/src/test/scala/spark/RDDSuite.scala:
```diff
@@ -173,4 +175,48 @@ class RDDSuite extends FunSuite with LocalSparkContext {
+    }.collect
+
+    assert(m2.toSet === (1 to 100).map{_ * 2}.toSet)
+    assert(acc2.value.keySet == (0 to 3).toSet)
```
This should be a triple-equals.

JoshRosen commented on core/src/test/scala/spark/RDDSuite.scala:
```diff
@@ -173,4 +175,48 @@ class RDDSuite extends FunSuite with LocalSparkContext {
+    assert(m2.toSet === (1 to 100).map{_ * 2}.toSet)
+    assert(acc2.value.keySet == (0 to 3).toSet)
+    //this condition will fail
+//    acc2.value.foreach { case(partition, values) =>
```
If this is supposed to fail, should we wrap it in an `intercept` block and check that an appropriate exception is actually thrown?

squito replied:
well, its not really _supposed_ to fail -- there is no exception, it just doesn't give the "expected" result.  that second part isn't really a unit test at all, its just documentation of why this method is needed.  probably doesn't belong here at all, I just wanted it as part of the pull request to demonstrate why the method was needed.

I guess I can just write it up somewhere else (seems too long to put in the spark docs also, at least in the current layout ...)

mateiz commented on core/src/main/scala/spark/RDD.scala:
```diff
@@ -34,6 +31,8 @@ import spark.rdd.ZippedRDD
 
 import SparkContext._
+import spark.RDD.PartitionMapper
+import util.CleanupIterator
```
Please don't use relative import names for packages (see http://spark-project.org/docs/latest/contributing-to-spark.html)

Commit:
use immutable map in test for clarity; fixup style issues

squito commented:
I cleaned up the style issues, sorry about that.
I left that "failing" test in there ... happy to remove it if you want me to, just wanted to be clear if you want me to document the trouble w/ mapPartitions somewhere

stephenh commented:
Hi Imran,

Have you seen the TaskContext and addOnCompleteCallback? That is what HadoopRDD uses to close the FileStream after all of the lines in a Hadoop file have been read.

You might be able to do what you're doing with just a custom RDD that did something like:

```
override def compute(s: Split, context: TaskContext): Iterator[(T, U)] = {
   setupDbConnection()
   context.addOnCompleteCallback { tearDownDbConnection()
   // call parent rdd or do own compute stuff
```

I believe this will achieve the same thing, as compute will be called on each partition, and you'll have start/stop hooks around the execution on each partition.

mateiz commented:
I agree with Stephen here. The addOnCompleteCallback mechanism also makes sure to call your handler if the task throws an exception, which is important.

Also, can you add a similar method in the Java API? I guess you would need to create an interface in Java for the setup/close stuff too.

Commit:
switch to using onCompleteCallback, to make sure cleanup is called even if exception is thrown

Commit:
fixup! switch

Commit:
remove CleanupIterator as its no longer used

Commit:
add java api for partition mapper

Commit:
more java api for partition mapper

squito commented:
good point, I definitely hadn't thought about ensuring cleanup w/ exceptions.

I've updated it to use onCompleteCallback.  I also added it to the java api -- I added separate classes for PairRDDs & DoubleRDDs, dunno if there is a better way to do that.

stephenh commented:
I wonder if this could be done with something more like decoration:

``` scala
val rdd = sc.textFile(...).setupPartitions(dbSetupLogic).mapPartitions(...).cleanupPartitions(dbCloseLogic)
```

So there would be two new RDDs, PartitionSetupRDD that first invoked its setup function once/partition, then called firstParent.compute, and then PartitionCleanupRDD, that setup the complete callback for its cleanup function.

Not sure if the decoupling would lead to unintended/nonsensical use cases. But, just musing, then perhaps they could be used separately, if you only need one or the other, or without the map, which PartitionMapper currently forces you do to.

Also, I just like that this would use plain functions and not a new "PartitionMapper" interface--for some reason that doesn't feel quite right, but I can't think of a better name.

I see what you're trying to do though.

JoshRosen commented on core/src/main/scala/spark/api/java/JavaRDDLike.scala:
```diff
@@ -117,6 +119,41 @@ trait JavaRDDLike[T, This <: JavaRDDLike[T, This]] extends PairFlatMapWorkaround
+   * setup & cleanup that happens before & after computing each partition
+   */
+  def mapWithSetupAndCleanup[K,V](m: JavaPairPartitionMapper[T,K,V]): JavaPairRDD[K,V] = {
+    val scalaMapper = new PartitionMapper[T,(K,V)] {
```
Can`JavaPairPartitionMapper<T, K, V>` be an abstract class that extends or implements `PartitionMapper<T, Tuple2<K, V>>`?  If you can do that, then you wouldn't have to wrap the the Java PartitionMapper to convert it into its Scala counterpart.

JoshRosen commented on core/src/main/scala/spark/api/java/ManifestHelper.java:
```diff
@@ -0,0 +1,11 @@
+import scala.reflect.ClassManifest;
+import scala.reflect.ClassManifest$;
+
+class ManifestHelper {
```
This ManifestHelper class is a good idea.  We could also use it to create the fake manifests in the Java `Function*` classes.

mateiz commented:
Regarding Stephen's comment -- I think it's better to keep PartitionMapper a single object instead of doing functions, in case you need to share state among the setup, map and clean methods (e.g. you open some external resource, use it in your map, then close it).

stephenh commented:
Ah, good point. That makes sense.

Commit:
JavaPairPartitionMapper directly implements PartitionMapper, so we can remove a helper object

squito commented:
I updated JavaPairPartitionMapper, per Josh's suggestion.  (We lose the ability for map to throw an exception, but that is already the case for the basic PartitionMapper.)

I tried doing the same thing for JavaDoubleRDD, but somehow I got stuck with weird manifest errors.  First it complained:

[error]  found   : ClassManifest[scala.Double]
[error]  required: ClassManifest[java.lang.Double]

Then when I switched to explicitly using a java.lang.Double manifest, it reversed:

[error]  found   : spark.RDD[java.lang.Double]
[error]  required: spark.RDD[scala.Double]

so I just left it as is

Commit:
cleanup java api; add some tests

squito commented:
OK I think this is ready to go now.  I got rid of the need for the helper object for JavaDoubleRDD, by just casting from java.lang.Double to scala.Double, and it seems happy.  Also I put a throws Exception declaration on map in PartitionMapper, for the java api.

squito commented:
Just curious what the status is on this -- waiting for some additional changes here, decided against merging it, or just haven't gotten to it yet.

mateiz commented on core/src/test/scala/spark/JavaAPISuite.java:
```diff
@@ -401,6 +398,54 @@ public void mapPartitions() {
   }
 
   @Test
+  public void mapPartitionsWithSetupAndCleanup() {
```
Please use two spaces for the indentation in this file (it looks like maybe it's tabs, or multiple spaces)

mateiz commented on core/src/main/scala/spark/RDD.scala:
```diff
@@ -340,6 +339,18 @@ abstract class RDD[T: ClassManifest](
+   * This might be useful if you need to setup some resources per task & cleanup them up at the end, eg.
+   * a db connection
+   */
+  def mapWithSetupAndCleanup[U: ClassManifest](
```
I would rename the instances of WithSetupAndCleanup to just WithCleanup for simplicity. Having cleanup will also imply that there's something to be cleaned up.

mateiz commented:
Sorry, just hadn't had a chance to look at it. It looks good but I made two small comments.

Commit:
Merge branch 'master' into map_setup_cleanup

Conflicts:
	core/src/main/scala/spark/RDD.scala

Commit:
rename to mapWithSetup; fix for split -> partition rename; fix spacing

squito commented:
thanks!  I've updated to take those comments into account.

Commit:
doh, fix missing refactor of mapWithSetup

squito commented:
wow, somhow I totally missed committing changes to one file before ... hope you didn't waste time looking at it before, now its actually all there

AmplabJenkins commented:
Can one of the admins verify this patch?

AmplabJenkins commented:
I'm the Jenkins test bot for the UC, Berkeley AMPLab. I've noticed your pull request and will test it once an admin authorizes me to. Thanks for your submission!

AmplabJenkins commented:
I'm the Jenkins test bot for the UC, Berkeley AMPLab. I've noticed your pull request and will test it once an admin authorizes me to. Thanks for your submission!

AmplabJenkins commented:
Thank you for your pull request. An admin will review this request soon.
